Updates to `ldap` module to improve user syncing in large environments

madmatt opened the pull request:
(Note: PR description updated 2019-01-16 to reflect all changes made in this PR)
Updates to `ldap` module to improve user syncing in large environments (#18)

# API Changes:
- Update `LDAPService` to always use the new `LDAPIterator` (see below) for calls to `LDAPService->getUsers()`. This ensures that all users will be returned, rather than only the first list of users (up to the LDAP maximum page size as was previously occurring).

- Deprecate memberLoggedIn() since it is now unused. Use afterMemberLoggedIn instead.


# Notable features:
## New way of syncing user and group data
Adds a new `LDAPIterator` class that provides the ability to return all objects that match a search filter, regardless of the LDAP servers max page size.

Note: There shouldn't be any breaking changes as the only option that we don't pass in is a search scope, and this is fine because the hard-coded scope (`Ldap::SEARCH_SCOPE_SUB`) ended up calling `ldap_search()` which the new iterator still does.

Note 2: Many thanks to GitHub user ThaDafinser for providing the initial implementation: https://gist.github.com/ThaDafinser/1d081bed8e5e6505e97bedf5863a187c

## Allow SilverStripe attributes to be reset by AD

Attributes missing from AD are not reset by the module, which can be a business case. This feature adds the ability for attributes to be reset if they are not present in the LDAP source data but present in the mapping.

This feature is opt-in as it's destructive (attributes missing from LDAP for a given user will be removed from the SilverStripe database).

## New method for photo thumbnail generation
Photos were not being saved previously, as the original module code was not updated to account for SilverStripe 4's new asset abstraction work. This has now been updated, along with some tweaks to ensure that the filesystem is not thrashed heavily.

LDAP sync works by syncing all users on a regular schedule (e.g. daily or hourly). Because LDAP does not natively support requesting 'only objects that have changed', we need to sync all users every time. This leads to thumbnail photos being marked as having changed every time the sync task runs for all users.

Instead of this, we now prevent this file version thrashing by comparing the binary blobs and only updating the file record if the files differ.

- Resolve issues with thumbnail storage (photos were not always being stored, with 'Extension is not allowed' errors in rare occurrences (~1 in 1,000), even though the extension is always set to .jpg)
- Prevent File ID thrashing, by updating the existing image object with the new image binary blob, instead of deleting the old File object and creating a new one (also by overwriting the old image instead of creating a new file, prevents `vN` from being appended to the end of the filename every time)
- Dont overwrite images that already exist already: When syncing large data sets, there's no point in overwriting images if it is the same image. Instead, check the file hash of the old and new, and overwrite if there's changes. For large LDAP sources, this resolves an issue where the syncs are incredibly file intensive.

## Resolve removed LDAP group mappings

Members which belonged to LDAP mapped groups weren't being removed from these associations if the LDAP data no longer contained the mappings. This meant if a group association was removed from LDAP the member was never updated.

This also accounts for situations where there's no more mappings or no mappings are given.

## Add extension hooks
Extension hooks are added for `onAfterLDAPMemberSyncTask` and `onAfterLDAPGroupSyncTask` to allow app-specific code to run after all members and groups are synced (for example, to map users by FK IDs instead of DN attribute values)


# Minor features:
- Add ability to get to the underlying `Zend\Ldap\Ldap` object to gain access to the PHP `ldap` resource

- Add new `BuildTask` that syncs just one LDAP user based on email address. This can be used to test LDAP syncing (make sure you run the `LDAPMigrateExistingMembersTask` task first if the user already exists in SS to map the GUID across)

- Properly save thumbnail photos (old code hadn't changed from the SS3 version and no longer worked with the new SS4 asset abstraction)

- Allow automatic update from LDAP when a member logs in to fail without breaking login based on configuration

- Bump minimum core version to 4.1


# Bug Fixes:
- Fix to use `afterMemberLoggedIn`, as the `memberLoggedIn` method is no longer called in core.
- Fix `LDAPDebugController::Users()` to not require so much memory by retrieving all attributes for all users (an issue now that we could return significantly more than 1,000 users)
- Set memory limit to unlimited in `LDAPDebugController`, allowing for massive resultsets
- Reset the pagination control after performing a paginated search, to allow additional searches to work
- Fix `LDAPAllSyncJob` to sync both groups and members, not just groups twice
- Set the max execution time to 60 minutes for user syncing and 15 minutes for group syncing, which can take a while under normal circumstances with large amounts of groups or users to sync
- Up memory limit to 1GB to account for massive datasets (LDAP doesn't support generators and pagination fails quickly as on busy LDAP servers pagination control cookies can get dropped from cache quickly, so we have to load all LDAP users into memory before processing any)
- Add `cn` in to the list of standard attributes to pull so we can use this in the log message
- Fix `LDAPMemberSyncOneTask` to fail quickly if no user is found in LDAP
- Fix minor security risk by not using the SAM Account Name (aka username) in the filename, instead using the user's GUID
- Update LDAPIterator::__construct docblock and constructor to support PHP5.6
- Fix issue with postgres compatibility: table identifiers not quoted
- Remove PHP 5 incompatible scalar type hinting
- Update config->get() style and remove unnecessary else clause, optimise imports
- Promote processSearchResults to protected
- Prefer getGateway() over direct property access, promote processThumbnailPhoto() visibility
- Use Injector to load a new Image instance
- Add public setter for LDAPService
- Use LDAPService setter, change die() for exit() and import missing Exception namespace
- Fix issue with postgres compatibility, unquoted column name
- Remove redundant limit(1) on DataList where first() is called afterwards
- Implement public setter for LDAPService in LDAPMemberSyncTask


# Testing: 
- Run automated phpcbf linting with slightly modified PSR-2 standard
- Fix code linting errors preventing successful build
- Add various core recipe versions to the Travis builds, add PHP 7.2 and 7.3
- Fix signature compatibility with DataObject::write() in SS 4.3 and remove ldap from Travis (installed by default)


# Documentation:
- Add documentation for using the new `LDAPIterator`
- Update developer documentation with corrected value for `groups_search_locations` YML fragment
- Add documentation for new features added to the module (`onAfterLDAPMemberSyncTask` and `onAfterLDAPGroupSyncTask`)
- Fix spelling mistakes in documentation
- Update comments to be technically correct (compared to the code they describe)

Commit:
Updates to `ldap` module to improve user syncing in large environments

API Changes:
- Update `LDAPService` to always use the new `LDAPIterator` (see below) for calls to `LDAPService->getUsers()`. This ensures that all users will be returned, rather than only the first list of users (up to the LDAP maximum page size as was previously occurring).

Features:
- Add new `LDAPIterator` class that provides the ability to return all objects that match a search filter, regardless of the LDAP servers max page size.

Note: There shouldn't be any breaking changes as the only option that we don't pass in is a search scope, and this is fine because the hard-coded scope (`Ldap::SEARCH_SCOPE_SUB`) ended up calling `ldap_search()` which the new iterator still does.

Note 2: Many thanks to GitHub user ThaDafinser for providing the initial implementation: https://gist.github.com/ThaDafinser/1d081bed8e5e6505e97bedf5863a187c

- Add ability to get to the underlying `Zend\Ldap\Ldap` object to gain access to the PHP `ldap` resource

- Add new `BuildTask` that syncs just one LDAP user based on email address. This can be used to test LDAP syncing (make sure you run the `LDAPMigrateExistingMembersTask` task first if the user already exists in SS to map the GUID across)

- [WIP] Properly save thumbnail photos (old code hadn't changed from the SS3 version and no longer works with the new SS4 asset abstraction)

- Add extension hooks for `onAfterLDAPMemberSyncTask` and `onAfterLDAPGroupSyncTask`

- Allow automatic update from LDAP when a member logs in to fail without breaking login based on configuration

Bug Fixes:
- Fix to use `afterMemberLoggedIn`, as the `memberLoggedIn` method is no longer called in core.
- Fix `LDAPDebugController::Users()` to not require so much memory by retrieving all attributes for all users (an issue now that we could return significantly more than 1,000 users)
- Set memory limit to unlimited in `LDAPDebugController`, allowing for massive resultsets
- Reset the pagination control after performing a paginated search, to allow additional searches to work
- Fix `LDAPAllSyncJob` to sync both groups and members, not just groups twice
- Fix thumbnail temp filenames
- Set the max execution time to 60 minutes for user syncing and 15 minutes for group syncing, which can take a while under normal circumstances with large amounts of groups or users to sync
- Up memory limit to 1GB to account for massive datasets (LDAP doesn't support generators and pagination fails quickly as on busy LDAP servers pagination control cookies can get dropped from cache quickly, so we have to load all LDAP users into memory before processing any)
- Add `cn` in to the list of standard attributes to pull so we can use this in the log message
- Fix `LDAPMemberSyncOneTask` to fail quickly if no user is found in LDAP

Documentation:
- Add documentation for using the new `LDAPIterator`
- Update developer documentation with corrected value for `groups_search_locations` YML fragment

Commit:
BUGFIX: Resolve issues with thumbnail storage

1) Fix minor security risk by not using the SAM Account Name (aka username) in the filename, instead using the user's GUID

2) Photos were not always being stored, with 'Extension is not allowed' errors in rare occurrences (~1 in 1,000), even though the extension is always set to .jpg

3) Prevents File ID thrashing, but updating the existing image object with the new image binary blog, instead of deleting the old File object and creating a new one.

4) Minor issue: By overwriting the old image instead of creating a new file, prevents `vN` from being appended to the end of the filename.

Commit:
Fix code linting errors preventing successful build

Commit:
Update LDAPIterator::__construct docblock and constructor to support PHP5.6

robbieaverill commented on src/Control/LDAPDebugController.php:
```diff
@@ -43,6 +43,7 @@ class LDAPDebugController extends ContentController
     protected function init()
     {
         parent::init();
+        ini_set('memory_limit', '-1'); // Increase memory limit to max possible
```
You could use `Environment::increaseMemoryLimitTo()` to re-use the core wrapper for this method

madmatt replied:
See comment below for this one: https://github.com/silverstripe/silverstripe-ldap/pull/18#discussion_r219886929

robbieaverill commented on src/Extensions/LDAPMemberExtension.php:
```diff
@@ -264,15 +282,22 @@ public function sync()
+                Injector::inst()->get(LDAPService::class)->updateMemberFromLDAP($this->owner);
+            } catch (Exception $e) {
+                // If the failure is acceptable, then ignore it and return. Otherwise, re-throw the exception
+                if ($this->owner->config()->allow_update_failure_during_login) {
```
Minor (stylistic thing): `$this->owner->config()->get('allow_update_failure_during_login')`

madmatt replied:
Have updated this now.

robbieaverill commented on src/Extensions/LDAPMemberExtension.php:
```diff
@@ -264,15 +282,22 @@ public function sync()
+                // If the failure is acceptable, then ignore it and return. Otherwise, re-throw the exception
+                if ($this->owner->config()->allow_update_failure_during_login) {
+                    return;
+                } else {
```
We could probably remove the else

madmatt replied:
Done!

robbieaverill commented on src/Iterators/LDAPIterator.php:
```diff
@@ -0,0 +1,294 @@
+
+        return $result;
+    }
+    private function getAttributeRecursive(string $dn, string $attrName, int $offset, int $maxPerRequest)
```
I'm not sure this module is PHP 7 minimum yet so we can't use scalar type hints

madmatt replied:
Sweet, have removed all these

robbieaverill replied:
I've unresolved a bunch of comments because they don't look to have been done, or maybe got reverted at some point =)

robbieaverill commented on src/Iterators/LDAPIterator.php:
```diff
@@ -0,0 +1,294 @@
+        if (! is_array($this->current)) {
+            $this->rewind();
+        }
+        if (! is_array($this->current)) {
```
There's quite a few style issues in this file I think, e.g. no spacing between methods. A phpcbf run with PSR-2 should fix them all though

madmatt replied:
Weird - I ran phpcs over this with the phpcs.xml.dist from framework and it didn't complain about this line. I've fixed the obvious ones at least.

robbieaverill commented on src/Model/LDAPGateway.php:
```diff
@@ -57,7 +80,17 @@ public function __construct()
+     * @param Iterator $records
+     * @return array
+     */
+    private function processSearchResults($records)
```
Could we use protected instead of private wherever possible? The iterator class is probably fine because its marked as final already

madmatt replied:
Yep no problem, that makes sense.

madmatt replied:
Will leave you to mark resolved when you're satisfied, sorry!

robbieaverill replied:
Still private

robbieaverill commented on src/Services/LDAPService.php:
```diff
@@ -355,7 +378,7 @@ public function getUsers($attributes = [])
 
         foreach ($searchLocations as $searchLocation) {
-            $records = $this->gateway->getUsers($searchLocation, Ldap::SEARCH_SCOPE_SUB, $attributes);
+            $records = $this->gateway->getUsersWithIterator($searchLocation, $attributes);
```
We could use `$this->getGateway()->...` here now

madmatt replied:
Yep good idea, have made that change throughout all of LDAPService

madmatt replied:
Will leave you to mark resolved when you're satisfied, sorry!

robbieaverill replied:
Still valid

robbieaverill commented on src/Services/LDAPService.php:
```diff
@@ -620,6 +607,65 @@ public function updateMemberFromLDAP(Member $member, $data = null, $updateGroups
+     * @param string $attributeName Name of the attribute in the $data array to get the binary blog from
+     * @return boolean true on success, false on failure
+     */
+    private function processThumbnailPhoto(Member $member, $fieldName, $data, $attributeName)
```
protected*

madmatt replied:
Done!

robbieaverill replied:
Still valid

robbieaverill commented on src/Services/LDAPService.php:
```diff
@@ -620,6 +607,65 @@ public function updateMemberFromLDAP(Member $member, $data = null, $updateGroups
+        if ($existingObj && $existingObj->exists()) {
+            $file = $existingObj;
+        } else {
+            $file = new Image();
```
```diff
- new Image();
+ Image::create();
```

madmatt replied:
Good catch, thanks - fixed!

robbieaverill replied:
Still valid

robbieaverill commented on src/Tasks/LDAPGroupSyncTask.php:
```diff
@@ -52,6 +57,8 @@ public function getTitle()
      */
     public function run($request)
     {
+        ini_set('max_execution_time', 900);
```
`Environment::increaseTimeLimitTo(900);`

madmatt replied:
Yeah so I tried that, but it doesn't work. `Environment::increaseTimeLimitTo()` and `Environment::increaseMemoryLimitTo()` increase only up to the maximum memory/time limit as defined (by default in php.ini).

So to do this 'properly' via Environment, I'd need to do:

```php
Environment::setMemoryLimitMax('1024M');
Environment::setTimeLimitMax(900);
Environment::increaseMemoryLimitTo();
Environment::increaseTimeLimitTo();
```

Which just seemed kinda overkill... but I can change that if you want?

robbieaverill replied:
Yeah that does seem like overkill. They're tasks so its probably fine to leave it the way it is =)

robbieaverill commented on src/Tasks/LDAPGroupSyncTask.php:
```diff
@@ -38,6 +38,11 @@ class LDAPGroupSyncTask extends BuildTask
+    /**
+     * @var LDAPService
+     */
+    public $ldapService;
```
I'd probably rather make this protected and have a public setter/getter

robbieaverill commented on src/Tasks/LDAPMemberSyncOneTask.php:
```diff
@@ -0,0 +1,90 @@
+        $user = $this->ldapService->getUserByEmail($email);
+
+        if (!$user) {
+            die(sprintf('No user found in LDAP for email %s', $email));
```
Maybe we could use `$this->log()` and then return instead of die? It's super minor but I know there won't be an EOL in this

phptek replied:
Yeah and if you _really_ want to stop PHP's execution, use `exit` instead, simply because from a dev's perspective, `die` looks like debug code that's been left-in!

robbieaverill replied:
I've updated it to use `exit` since there's an existing example of the same kind of direct output above. It's OK for now!

robbieaverill commented on src/Tasks/LDAPMemberSyncTask.php:
```diff
@@ -54,11 +59,14 @@ public function getTitle()
     public function run($request)
     {
+        ini_set('max_execution_time', 3600); // 3600s = 1hr
+        ini_set('memory_limit', '1024M'); // 1GB memory limit
```
```php
Environment::increaseTimeLimitTo(3600);
Environment::increaseMemoryLimitTo('1024M');
```

robbieaverill replied:
Ok to leave this

robbieaverill commented on src/Tasks/LDAPMemberSyncTask.php:
```diff
@@ -149,4 +157,23 @@ protected function log($message)
+     */
+    protected function findOrCreateMember($data = [])
+    {
+        $member = Member::get()->filter('GUID', $data['objectguid'])->limit(1)->first();
```
I think that the `->limit(1)` part of this is redundant

robbieaverill reviewed (changes requested):
Some minor suggestions =)

Commit:
Feature - Allow SilverStripe attributes to be reset by AD

Attributes missing from AD are not reset by the module, which can
be a business case. This feature adds the ability for attributes
to be reset if they are not present in the LDAP source data but
present in the mapping.

This feature is opt-in as it's destructive.

Commit:
Dont overwrite images that already exist already

When syncing large data sets, there's no point in overwriting images if
it is the same image. Instead, check the file hash of the old and new,
and overwrite if there's changes.

For large LDAP sources, this resolves an issue where the syncs are
incredibly file intensive.

Commit:
Resolve removed LDAP group mappings

Members which belonged to LDAP mapped groups weren't being removed from
these associations if the LDAP data no longer contained the mappings.

This meant if a group association was removed from LDAP the member was
never updated.

This also must Account for situations where there's no more mappings
or no mappings are given.

Commit:
fix spelling mistakes in documentation

Commit:
Make sure functions exist before calling them.

Commit:
ensure comments are technically correct.

madmatt committed:
Merge pull request #3 from caffeineinc/bugfix/fix-remove-group-mappings

Fix group mappings for removed Groups

robbieaverill commented on src/Extensions/LDAPMemberExtension.php:
```diff
@@ -264,15 +283,22 @@ public function sync()
+     * this will update the Member record from LDAP data.
      */
-    public function memberLoggedIn()
+    public function afterMemberLoggedIn()
```
I guess this is technically a semver break, even if the old method name was wrong to start with. Maybe we could keep the old one and `return $this->afterMemberLoggedIn()` with a deprecation notice on it?

madmatt replied:
Thanks for doing this @robbieaverill, I missed that when upgrading!

robbieaverill reviewed (changes requested):
Re-reviewed, looks good overall, there's still some unresolved comments from earlier and the tests need to be fixed. I'll have a quick look at this now.

robbieaverill committed:
Run automated phpcbf linting with slightly modified PSR-2 standard

robbieaverill committed:
Add various core recipe versions to the Travis builds, add PHP 7.2 and 7.3

robbieaverill committed:
Bump minimum core version to 4.1

robbieaverill committed:
Fix signature compatibility with DataObject::write() in SS 4.3 and remove ldap from Travis (installed by default)

robbieaverill committed:
Fix issue with postgres compatibility: table identifiers not quoted

robbieaverill committed:
FIX Remove PHP 5 incompatible scalar type hinting

robbieaverill committed:
Update config->get() style and remove unnecessary else clause, optimise imports

robbieaverill committed:
Promote processSearchResults to protected

robbieaverill committed:
Prefer getGateway() over direct property access, promote processThumbnailPhoto() visibility

robbieaverill committed:
FIX Use Injector to load a new Image instance

robbieaverill committed:
Add public setter for LDAPService

robbieaverill committed:
Use LDAPService setter, change die() for exit() and import missing Exception namespace

robbieaverill committed:
Fix issue with postgres compatibility, unquoted column name

robbieaverill committed:
API Deprecate memberLoggedIn() since it is now unused. Use afterMemberLoggedIn instead.

robbieaverill committed:
Remove redundant limit(1) on DataList where first() is called afterwards

robbieaverill committed:
FIX Implement public setter for LDAPService in LDAPMemberSyncTask

robbieaverill reviewed (approved):
I've done some polishing, tests are green now and I've addressed my feedback 😆If you're happy @madmatt then feel free to merge

madmatt commented on src/Tasks/LDAPMemberSyncTask.php:
```diff
@@ -28,7 +28,7 @@ class LDAPMemberSyncTask extends BuildTask
      */
     private static $dependencies = [
-        'ldapService' => '%$' . LDAPService::class,
+        'LDAPService' => '%$' . LDAPService::class,
```
This doesn't match the actual `protected $ldapService` and usage within the class (`$this->ldapService->x()` - shouldn't we change this back to be `'ldapService' => xxx`?

robbieaverill replied:
It needs to match the setter or the public property (whichever exists) eg (set)LDAPService()

madmatt replied:
Ahh gotcha, okay I see - so it matches the `setLDAPService()` method which in turn sets `$this->ldapService`, which is then used in the class. Okay, that's fine by me. Thanks!

madmatt commented:
Thanks @robbieaverill, looks great! Just one question/comment remaining from me, then I can squash and you can merge if that makes sense?

It's a bit unwieldy now, here's the comment: https://github.com/silverstripe/silverstripe-ldap/pull/18/files#r248058868 (or it's just above this comment too)

madmatt commented:
Not sure why Scrutinizer says it's still running, clicking through shows it as having completed and passed. Happy with the Codecov patch failures, we need to an automated test pass on this module but it shouldn't hold up this work IMO.

* Original development done by @madmatt
* Reviewed by @robbieaverill, updates made
* Updates made reviewed by @madmatt

Thanks for the assist @robbieaverill, merging this in now. Shall we release a new major version once this is up? Or maybe just a minor release? Theoretically I don't think there's any breaking changes, but I can't be sure so would probably prefer to bump the major version just in case.

ScopeyNZ commented:
It looks like @robbieaverill intended a minor and made code changes to that effect. I think we're fine tagging a minor for this. Would you like me to do this for you?
